### PR TITLE
Add implicit `safeFree` for `RapidsBuffer`

### LINKED ISSUE #5098: [BUG] Harden calls to `RapidsBuffer.free`
As @gerashegalov mentioned in https://github.com/NVIDIA/spark-rapids/pull/5093#pullrequestreview-926548173, there are various places in the code where `free` is called on a `RapidsBuffer` instance with the assumption that it does not throw.  We either should have something like a `safeFree` implicit utility function or appropriate `try...finally` blocks to ensure attempts to release other resources at the same time are not interrupted.

### PR DESCRIPTION
**Pull request by HaoYang670**
Signed-off-by: remzi <13716567376yh@gmail.com>

Close #5098 

Add implicit classes to safely free `RapidsBuffer`s

### CODE REVIEW

**Comment by HaoYang670** [comment]
Not add tests yet.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/implicits.scala [hunk 1]
```diff
@@ -59,6 +59,30 @@ object RapidsPluginImplicits {
     }
   }
 
+  implicit class RapidsBufferColumn[A <: RapidsBuffer](rapidsBuffer: RapidsBuffer) {
```
Comment (firestarman) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/implicits.scala:62]:
```suggestion
  implicit class RapidsBufferColumn[A <: RapidsBuffer](rapidsBuffer: A) {
```
??

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/implicits.scala [hunk 2]
```diff
@@ -59,6 +59,30 @@ object RapidsPluginImplicits {
...
+    def safeFree(e: Throwable = null): Unit = {
+      if (rapidsBuffer != null) {
+        if (e != null) {
+          try {
+            rapidsBuffer.free()
+          } catch {
+            case suppressed: Throwable => e.addSuppressed(suppressed)
+          }
+        } else {
+          rapidsBuffer.free()
```
Comment (sperlingxx) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/implicits.scala:80]: Why do we free a null ptr?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/implicits.scala [hunk 3]
```diff
@@ -59,6 +59,30 @@ object RapidsPluginImplicits {
...
   }
 
+  implicit class RapidsBufferColumn[A <: RapidsBuffer](rapidsBuffer: RapidsBuffer) {
+
+    /**
+     * safeFree: Is an implicit on RapidsBuffer class that tries to free the resource, if an
+     * Exception was thrown prior to this free, it adds the new exception to the suppressed
+     * exceptions, otherwise just throws
+     *
+     * @param e Exception which we don't want to suppress
```
Comment (sperlingxx) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/implicits.scala:69]: Does the comment align with the implementation ?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala [hunk 4]
```diff
@@ -145,7 +146,7 @@ class RapidsBufferCatalog extends Logging {
   def removeBuffer(id: RapidsBufferId): Unit = {
     val buffers = bufferMap.remove(id)
     if (buffers != null) {
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:148]: do we need this null check? It seems the implicit checks that the sequence is not null, so we should be really safe to just do:
```
val buffers = bufferMap.remove(id)
buffers.safeFree()
```

**Comment by HaoYang670** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/implicits.scala [hunk 5]
```diff
@@ -87,11 +111,44 @@ object RapidsPluginImplicits {
...
   }
 
+  implicit class RapidsBufferSeq[A <: RapidsBuffer](val in: SeqLike[A, _]) {
+    /**
+     * safeFree: Is an implicit on a sequence of RapidsBuffer classes that tries to free each
+     * element of the sequence, even if prior free calls fail. In case of failure in any of the
+     * free calls, an Exception is thrown containing the suppressed exceptions (getSuppressed),
+     * if any.
+     */
+    def safeFree(error: Throwable = null): Unit = if (in != null) {
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/implicits.scala:121]: looks like significant duplication between safeFree and safeCLose? Can you try to extract common logic into a common base class, and maybe use a lambda `fn: A => Unit` to implement the difference between close and free?

**Comment by HaoYang670** [comment]
build

**Comment by HaoYang670** [comment]
Close this PR as something weird happened when merging the branch. I will open a new PR.

**Comment by HaoYang670** [comment]
open #5471

**Pull request closed without merging**